### PR TITLE
Heron DSL documentation

### PR DESCRIPTION
**Pull request by lucperkins**
While we're waiting on stable releases for the Java and Python DSLs, I'd like to first add this more theoretical documentation explaining the programming model, benefits, comparison with the topology API, etc.

@srkukarni @huijunw Please note that this is a WIP.

### CODE REVIEW

File: website/content/docs/concepts/topologies.md [hunk 1]
```diff
@@ -2,6 +2,15 @@
 title: Heron Topologies
 ---
 
+> ## New DSL for Heron
+> As of Heron version 0.15.2, there is a new **Heron DSL** that can be used
```
Comment (srkukarni) [inline website/content/docs/concepts/topologies.md]: This should be 0.16.0

File: website/content/docs/concepts/topologies.md [hunk 2]
```diff
@@ -42,6 +51,156 @@ lifecycle of a topology, which typically goes through the following stages:
...
+streamlets*. This can be done using a wide variety of available operations, including
+many you may be familiar with from functional programming:
+
+Operation | Description
+:---------|:-----------
+map | Returns a new streamlet by applying the supplied mapping function to each element in the original streamlet
+flatMap | Like a map operation but with the important difference that each element of the streamlet is flattened
+join | Joins two separate streamlets into a single streamlet
+filter | Returns a new streamlet containing only the elements that satisfy the supplied filtering function
+sample | Returns a new streamlet containing only a fraction of elements. That fraction is defined by the supplied function.
```
Comment (srkukarni) [inline website/content/docs/concepts/topologies.md]: this is not supported in dsl. Might be removed from python dsl as well

File: website/content/docs/concepts/topologies.md [hunk 3]
```diff
@@ -42,6 +51,156 @@ lifecycle of a topology, which typically goes through the following stages:
...
+import com.twitter.heron.dsl.impl.BaseStreamlet;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+public final class ExampleDSLTopology {
+    public ExampleDSLTopology() {}
+
+    public static void main(String[] args) {
+        Streamlet<Integer> randoms = BaseStreamlet.createSupplierStreamlet(() -> ThreadLocalRandom.current().nextInt(1, 11));
+        Streamlet<Integer> zeroes = BaseStreamlet.createSupplierStreamlet(() -> 0);
```
Comment (srkukarni) [inline website/content/docs/concepts/topologies.md]: createSupplierStreamlets are now not public. Please use Builder.newSource

File: website/content/docs/concepts/topologies.md [hunk 4]
```diff
@@ -42,6 +51,156 @@ lifecycle of a topology, which typically goes through the following stages:
...
+lifecycles](#topology-lifecycle), the resulting topologies are equivalent. From a
+development workflow standpoint, however, the difference is profound. The Heron
+DSL allows for topology code that is:
+
+* Less verbose and far less dependent on boilerplate
+* 
+
+### Streamlets
+
+The core construct underlying the Heron DSL is that of the **streamlet**. A streamlet is
```
Comment (huijunw) [inline website/content/docs/concepts/topologies.md]: what is difference between DSL streamlet and existing data model stream?

File: website/content/docs/concepts/topologies.md (hunk 1 above)
Comment (lucperkins) [reply website/content/docs/concepts/topologies.md]: Please note that this is a work in progress

File: website/content/docs/concepts/topologies.md [hunk 5]
```diff
@@ -42,6 +51,156 @@ lifecycle of a topology, which typically goes through the following stages:
...
+1. RAM
+
+The defaults:
+
+Resource | Default
+:--------|:-------
+Number of containers | 1
+CPU | 1.0
+RAM | 512 MB
+
```
Comment (srkukarni) [inline website/content/docs/concepts/topologies.md]: Can we talk about how Resources can be specified?

Comment (lucperkins) [reply website/content/docs/concepts/topologies.md]: Again, this is a work in progress. I'll ping you as soon as it's in reviewable shape.

**Comment by lucperkins** [comment]
@srkukarni This is ready for review. Unfortunately this needs to be a little bit quick and dirty while I finish up the Java DSL docs. But it'd be nice to get this out there to pass on to interested parties.

File: website/content/docs/concepts/topologies.md [hunk 6]
```diff
@@ -2,45 +2,295 @@
 title: Heron Topologies
 ---
 
+> ## New Functional API for Heron
+> As of version 0.15.2, Heron offers a new **Functional API** that you can use
```
Comment (srkukarni) [inline website/content/docs/concepts/topologies.md]: 0.16.0

Comment (lucperkins) [reply website/content/docs/concepts/topologies.md]: Got it

File: website/content/docs/concepts/topologies.md [hunk 7]
```diff
@@ -2,45 +2,295 @@
...
+
+When Heron was first created, the model for creating topologies was deeply
+indebted to the Apache Storm model. Under that model, developers creating topologies
+needed to explicitly define the behavior of every spout and bolt in the topology.
+Although this provided a powerful low-level API for creating topologies, that approach
+presented a variety of drawbacks for developers:
+
+* **Verbosity** --- In both the Java and Python topology APIs, creating spouts and bolts involved substantial boilerplate, requiring developers to both provide implementations for spout and bolt classes and also specify the connections between those spouts and bolts. This often led to the problem of...
+* **Difficult debugging** --- When spouts, bolts, and the connections between them need to be created "by hand," a great deal of cognitive load
+* **Tuple-based data model** --- In the older topology API, spouts and bolts passed tuples and nothing but tuples within topologies. Although tuples are a powerful and flexible data type, the topology API forced *all* spouts and bolts to serialize or deserialize tuples.
```
Comment (srkukarni) [inline website/content/docs/concepts/topologies.md]: Can we include Type safety

Comment (lucperkins) [reply website/content/docs/concepts/topologies.md]: Yep

**Pull request merged**